caikit/core/model_management/model_trainer_base.py

```diff
@@ -0,0 +1,88 @@
+# Copyright The Caikit Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+A Trainer is responsible for managing execution of a training job for a given
+module class
+"""
+
+# Standard
+from enum import Enum
+from typing import Optional, Type
+import abc
+
+# Local
+from ..modules import ModuleBase
+from ..toolkit.factory import FactoryConstructible
+
+
+class ModelTrainerBase(FactoryConstructible):
+    __doc__ = __doc__
+
+    class TrainingStatus(Enum):
+        """A given training job must be in exactly one of these states at all
+        times
+        """
+
+        QUEUED = 1
+        RUNNING = 2
+        COMPLETED = 3
+        CANCELED = 4
+        ERRORED = 5
+
+    class ModelFutureBase(abc.ABC):
+        """Every Trainer must implement a ModelFuture class that can access the
+        training job in the infrastructure managed by the trainer.
+        """
+
+        @property
+        @abc.abstractmethod
+        def id(self) -> str:
+            """Every model future must have a unique ID that can be used to look
+            up the in-flight training
+            """
+
+        @property
+        @abc.abstractmethod
+        def save_path(self) -> Optional[str]:
+            """If created with a save path, the future must expose it"""
+
+        @abc.abstractmethod
+        def get_status(self) -> "TrainingStatus":
+            """Every model future must be able to poll the status of the
+            training job
+            """
+
+        @abc.abstractmethod
+        def cancel(self):
+            """Terminate the given training"""
+
+        @abc.abstractmethod
+        def wait(self):
+            """Block until the job reaches a terminal state"""
+
+        @abc.abstractmethod
+        def load(self) -> ModuleBase:
+            """A model future must be loadable with no additional arguments"""
+
+    @abc.abstractmethod
+    def train(
```

Comment:
design question, since its returning a future, would it make sense to have it all around python `asyncio` (which also has semantically similar API, where it returns "future" and on a future, we can do "wait" etc ?

Reply:
There's _definitely_ a world in which we move the entire `train` ecosystem to `asyncio`, but my worry about doing that is that it could enforce a lot of complexity on the author of each `module` to write `async def train` correctly. The whole goal of the `module` interface is to allow users to not think too hard about how the body of the function will be executed, and I feel like requiring `asyncio` would be a pretty big lift for a lot of folks.

If we didn't require the individual `train` functions to be defined as async functions, then the framework would need to have some very fancy footwork to map from the top-level `async` function to the underlying synchronous function. It would also mean that the top-level `caikit.train` would logically want to be `async` which would be quite a lot harder for folks to "just call."

I'll also point out that `concurrent.futures` liberally uses the `futures` pattern with alternate concurrency mechanisms (`ThreadPoolExecutor`, `ProcessPoolExecutor`), so `future` is not "just for async" in python.

Reply:
I'm going to punt on the `asyncio` question since that would be a major refactor of a lot of the architecture

Reply:
Yep, asyncio could bring in some interesting implications on how we wrap the train function in runtime 🤔